### PR TITLE
Add:  linux time wheel

### PR DESCRIPTION
**Pull request by AlexStocks**
**What this PR does**:

Add:
1 linux time wheel

### CODE REVIEW

File: time/timer.go [hunk 1]
```diff
@@ -0,0 +1,649 @@
...
+	uatomic "go.uber.org/atomic"
+)
+
+var (
+	ErrTimeChannelFull   = fmt.Errorf("timer channel full")
+	ErrTimeChannelClosed = fmt.Errorf("timer channel closed")
+)
+
+// Init initializes a default timer wheel
+func Init() {
```
Comment (watermelo) [inline time/timer.go:40]: Is it `init`? otherwise the defaultTimerWheel cannot be initialized by default.

File: time/timer.go [hunk 2]
```diff
@@ -0,0 +1,649 @@
...
+	maxMinute = 60
+	maxHour   = 24
+	maxDay    = 31
+	// ticker interval不能设置到这种精度，
+	// 实际运行时ticker的时间间隔会在1.001ms上下浮动,
+	// 当ticker interval小于1ms的时候，会导致TimerWheel.hand
+	// 和timeWheel.inc不增长，造成时间错乱：例如本来
+	// 1.5s运行的函数在持续2.1s之后才被执行
+	// minDiff = 1.001 * MS
+	minDiff       = 10e6
```
Comment (wongoo) [inline time/timer.go:74]: minTickerInterval

File: time/timer.go [hunk 3]
```diff
@@ -0,0 +1,649 @@
...
+	curGxTime             = time.Now().UnixNano() // current goext time in nanoseconds
+)
+
+const (
+	maxMS     = 1000
+	maxSecond = 60
+	maxMinute = 60
+	maxHour   = 24
+	maxDay    = 31
+	// ticker interval不能设置到这种精度，
```
Comment (wongoo) [inline time/timer.go:68]: change to english

File: time/timer.go [hunk 4]
```diff
@@ -0,0 +1,649 @@
...
+// TimerType defines a timer task type.
+type TimerType int32
+
+const (
+	eTimerOnce TimerType = 0x1 << 0
+	eTimerLoop TimerType = 0x1 << 1
+)
+
+// AddTimer returns a timer struct obj.
+// 异步通知timerWheel添加一个timer，有可能失败
```
Comment (wongoo) [inline time/timer.go:501]: change to english

File: time/timer.go [hunk 5]
```diff
@@ -0,0 +1,649 @@
...
+func dayNum(expire int64) int64    { return expire / (maxHour * int64(time.Hour)) }
+
+// TimerFunc defines the time func.
+// if the return error is not nil, the related timer will be closed.
+type TimerFunc func(ID TimerID, expire time.Time, arg interface{}) error
+
+// TimerID is the id of a timer node
+type TimerID = uint64
+
+type timerNode struct {
```
Comment (wongoo) [inline time/timer.go:91]: pls add comment for this struct and its fields, so that to get its functionality

File: time/timer.go (hunk 1 above)
Comment (AlexStocks) [reply time/timer.go:40]: Yeah, I supply this func to let user init defaultTimerWheel as his will.

File: time/timer.go [hunk 6]
```diff
@@ -0,0 +1,649 @@
...
+
+			case nodeAction, qFlag = <-w.timerQ:
+				// 此处只用一个channel，保证对同一个timer操作的顺序性
+				if qFlag {
+					switch {
+					case nodeAction.action == ADD_TIMER:
+						w.number.Add(1)
+						w.insertTimerNode(nodeAction.node)
+					case nodeAction.action == DEL_TIMER:
+						w.number.Add(1)
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply time/timer.go:223]: Yeah, u r right.

File: time/timer.go [hunk 7]
```diff
@@ -0,0 +1,649 @@
...
+	slot = w.slot[0]
+	clock = atomic.LoadInt64(&w.clock)
+	var next *list.Element
+	for e := slot.Front(); e != nil; e = next {
+		node = e.Value.(timerNode)
+		if clock < node.trig {
+			break
+		}
+
+		err = node.timerRun(node.ID, UnixNano2Time(clock), node.arg)
```
[the remark this answers is not in the document]
Comment (AlexStocks) [reply time/timer.go:280]: 用户自己可以在 timeRun 里面启动 gr 异步执行他的逻辑

File: time/timer.go (hunk 2 above)
Comment (AlexStocks) [reply time/timer.go:74]: done

File: time/timer.go (hunk 3 above)
Comment (AlexStocks) [reply time/timer.go:68]: this var is private local var. what's more, my english is not so good that i can translate it into accurate english text to let user know what I mean. maybe u can help me.

File: time/timer.go (hunk 5 above)
Comment (AlexStocks) [reply time/timer.go:91]: done

File: time/timer.go (hunk 4 above)
Comment (AlexStocks) [reply time/timer.go]: done

File: time/sleep_test.go [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+	timer = AfterFunc(TimeSecondDuration(4.5), f)
+	// 添加是异步进行的，所以sleep一段时间再去检测timer number
+	time.Sleep(1e9)
+	assert.Equalf(t, 1, defaultTimerWheel.TimerNumber(), "before stop")
+	timer.Stop()
+	// 删除是异步进行的，所以sleep一段时间再去检测timer number
+	time.Sleep(1e9)
+
+	time.Sleep(TimeSecondDuration(0.01))
+	//assert.Equalf(t, 0, defaultTimerWheel.TimerNumber(), "after stop")
```
Comment (wongoo) [inline time/sleep_test.go:177]: should assert after stop?

**Comment by wongoo** [comment]
@AlexStocks  how about open a online meeting to explain the code for reviewers?

File: time/sleep.go [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+// expired or been stopped.
+func (t *Timer) Reset(d time.Duration) {
+	if d <= 0 {
+		return
+	}
+	if t.w == nil {
+		panic("time: Stop called on uninitialized Timer")
+	}
+
+	t.w.resetTimer(t, d)
```
Comment (Patrick0308) [inline time/sleep.go:84]: pls handle err

File: time/sleep.go [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+}
+
+// Stop prevents the Timer from firing.
+func (t *Timer) Stop() {
+	if t.w == nil {
+		panic("time: Stop called on uninitialized Timer")
+	}
+
+	t.w.deleteTimer(t)
```
Comment (Patrick0308) [inline time/sleep.go:93]: as above

**Review by Patrick0308 — COMMENTED** [review]
Why can not see time wheel in `Files changed`?

File: time/sleep.go (hunk 9 above)
Comment (AlexStocks) [reply time/sleep.go:84]: done

File: time/sleep.go (hunk 10 above)
Comment (AlexStocks) [reply time/sleep.go:93]: done

File: time/sleep_test.go (hunk 8 above)
Comment (AlexStocks) [reply time/sleep_test.go:177]: 并行跑所有的 test 的情况下，这行有可能失败。所以注释掉了

File: time/timer.go (hunk 3 above)
Comment (AlexStocks) [reply time/timer.go]: done

**Review by wongoo — APPROVED** [review]

**Review by georgehao — APPROVED** [review]

**Review by watermelo — APPROVED** [review]

**Pull request merged**